osf/models/mixins.py

```diff
@@ -739,6 +739,52 @@ def subject_hierarchy(self):
             ]
         return []
 
+    def check_subject_perms(self, auth):
+        AbstractNode = apps.get_model('osf.AbstractNode')
+        Preprint = apps.get_model('osf.Preprint')
+        CollectionSubmission = apps.get_model('osf.CollectionSubmission')
+
+        if isinstance(self, (AbstractNode, Preprint)):
+            if not self.has_permission(auth.user, ADMIN):
+                raise PermissionsError('Only admins can change subjects.')
+        elif isinstance(self, CollectionSubmission):
+            if not self.guid.referent.has_permission(auth.user, ADMIN) and not auth.user.has_perms(self.collection.groups[ADMIN], self.collection):
+                raise PermissionsError('Only admins can change subjects.')
+        return
+
+    def add_subjects_log(self, old_subjects, auth):
+        self.add_log(
+            action=NodeLog.SUBJECTS_UPDATED,
+            params={
+                'subjects': list(self.subjects.values('_id', 'text')),
+                'old_subjects': list(Subject.objects.filter(id__in=old_subjects).values('_id', 'text'))
+            },
+            auth=auth,
+            save=False,
+        )
+        return
+
+    def set_subjects_from_relationships(self, subjects, auth, add_log=True):
+        """ Helper for setting M2M subjects field from list of flattened subjects received from UI.
+        Only authorized admins may set subjects.
+
+        :param list[Subject._id] new_subjects: List of flattened subject hierarchies
+        :param Auth auth: Auth object for requesting user
+        :param bool add_log: Whether or not to add a log (if called on a Loggable object)
+
+        :return: None
+        """
+        self.check_subject_perms(auth)
+        old_subjects = list(self.subjects.values_list('id', flat=True))
+        self.subjects.clear()
+        for subj in validate_subjects(subjects):
```

Comment:
Main undecided questions:

What should we expect subjects request to look like as relationships, and what should we store in the db?
1. Current state: we send in hierarchies as a list of lists under `attributes`. For example, assume hierarchy Fruits > Apples,  Fruits > Oranges, currently we send in with [[Fruits, Apples], [Fruits, Oranges]].  3 subjects go in db, Fruits, Apples, Oranges
2. Future state:
    1. Under relationships (with proper JSONAPI), send in [Fruits, Apples, Oranges], or [Apples, Oranges]?
    2. What are we validating? Should we expect all parts of the hierarchy to show up in the relationships, or should we just expect highest subjects in hierarchy? Should we just store highest level, or should we construct hierarchy ourselves and store what we store now?

Reply:
Making sure I understand your "current state" example:
1. user selects `Apples` and `Oranges`
2. frontend sends to API: `subjects: [[Fruits, Apples], [Fruits, Oranges]]`
3. backend saves to DB: `subjects: [Fruits, Apples, Oranges]`
4. API now shows `subjects: [[Fruits, Apples], [Fruits, Oranges]]`, and `Fruits` is left implicit (?)

There are multiple ways we could approach this in JSON:API, but what I think makes the most sense is:
- when updating, allow `subjects: [Apples, Oranges]` and implicitly add `Fruits` as well. sending `subjects: [Fruits, Apples, Oranges]` should also be acceptable.
- when listing subjects for a node/reg/preprint, include all parts of the hierarchy -- no matter how the above were sent, get `subjects: [Fruits, Apples, Oranges]`
- when serializing a subject, include a `parent` relationship (or `parent_subject`? or some other appropriate name)
  - `Fruits` has `parent: null`
  - `Apples` has `parent: Fruits`
  - since they're all included when we fetch `subjects`, we can render it using `subject.parent` recursively and take advantage of front-end caching

Unless... can a subject have multiple parents?

Reply:
Your understanding of the current state is correct @aaxelb!

Your recommendations of future state is similar to how I've started to build it out, except for what I'm storing in the db - I'm currently allowing any subjects in request, not enforcing the whole hierarchy to be there, but then I'm only storing those specific subjects.  Sending in [Apples, Oranges] and only [Apples, Oranges] stored. 

So in this PR, when serializing a registration's subjects, I'm only returning Apples, for example, if I just sent in apples. I'm using existing taxonomy serializer, which already has parents links, but they're apparently under "links" not "relationships": 

```JSON
GET http://localhost:8000/v2/registrations/p7dyu/subjects/
{
    "data": [
        {
            "links": {
                "self": "http://localhost:8000/v2/taxonomies/5ae21b3c195426000d952b36/",
                "parents": [
                    "http://localhost:8000/v2/taxonomies/5ae21b3a195426000d952b05/"
                ]
            },
            "attributes": {
                "text": "2D Materials",
                "path": "ECSarXiv|Engineering|Electronic Materials and Processing|2D Materials",
                "child_count": 0,
                "parent": {
                    "text": "Electronic Materials and Processing",
                    "id": "5ae21b3a195426000d952b05"
                },
                "share_title": "ECSarXiv"
            },
            "type": "taxonomies",
            "id": "5ae21b3c195426000d952b36"
        }
    ],
    "meta": {
        "total": 1,
        "per_page": 10,
        "version": "2.15"
    },
    "links": {
        "self": "http://localhost:8000/v2/registrations/p7dyu/subjects/",
        "first": null,
        "last": null,
        "prev": null,
        "next": null
    }
}
```
So sounds like I instead need to store [Fruits, Apples, Oranges] in the db, regardless of whether [Apples, Oranges] or [Fruits, Apples, Oranges] were sent in.  On request, build hieararchy, and store all those subjects.  This will make the request to fetch the subjects easy, just pull whatever's in the db.  And I might need to add a parents relationship to the existing TaxonomySerializer since that parents link might not work.  

I don't know if subjects can have more than one parent, I'll find out.

Reply:
Huh, it has a `parent` attribute, `parents` links, *and* `path` attribute. How thorough.

> On request, build hierarchy, and store all those subjects.

^ that sounds right to me! And that matches the existing behavior, right? So no need to change the old logic or migrate anything?

There are a few things that would ideally change about that serialization... not sure if they're all worth doing right now, though:
- remove `parent` attr, `parents` links (replaced by `parent` relationship)
- replace `child_count` attr with a `children` relationship that supports `related_counts` (so we're not running `self.children.count()` every time a subject is serialized)
- `type: "subjects" instead of `type: "taxonomies"`

Reply:
@aaxelb, yes, it matches existing behavior regarding what is stored, so no migration needed, but logic will have to be added to save it that way since the request body is different. 

I'm concerned about the implications for modifying existing TaxonomySerializer since I assume preprints is using it.  Who is responsible for preprints/collections front-end changes?  I assume I'd have to version all of this as well, with similar API magic that we're putting in for subjects, since we'd have both a parent attribute/relationship we'd have to maintain.

Reply:
I was describing my ideal (aka JSON:API compliant and semantically correct), but we might be better off finding compromise if it's too much bother.

But what if, instead of modifying `TaxonomySerializer`, we add a `SubjectSerializer` and a new `/v2/subjects/`? The existing`/v2/taxonomies/` could be deprecated and the existing `subjects` attribute on nodes/preprints/etc omitted if `version >= X`. Would that simplify things, or just add more mess?

I believe @jamescdavis would be responsible for updating preprints front end, but preferably versioning would save him from having to.

Reply:
I understand - I too want to get everything nice and json-api compliant, but it's going to take a lot of work to get it there!

A `SubjectSerializer` sounds like a nice compromise.